File: app/src/main/java/com/hyapp/achat/view/adapter/MessageAdapter.java
package com.hyapp.achat.view.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.text.format.DateUtils;
import android.util.Pair;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;

import com.aghajari.rlottie.AXrLottieImageView;
import com.facebook.drawee.view.SimpleDraweeView;
import com.hyapp.achat.R;
import com.hyapp.achat.model.entity.People;
import com.hyapp.achat.model.entity.SortedList;
import com.hyapp.achat.viewmodel.utils.TimeUtils;
import com.hyapp.achat.model.entity.ChatMessage;
import com.hyapp.achat.model.entity.Contact;
import com.hyapp.achat.model.entity.DetailsMessage;
import com.hyapp.achat.model.entity.LottieMessage;
import com.hyapp.achat.model.entity.Message;
import com.hyapp.achat.model.entity.ProfileMessage;
import com.hyapp.achat.model.entity.TextMessage;
import com.hyapp.achat.model.entity.utils.MessageUtils;
import com.hyapp.achat.model.entity.utils.PersonUtils;
import com.hyapp.achat.view.component.emojiview.view.AXEmojiTextView;
import com.hyapp.achat.view.component.GroupAvatarView;
import com.hyapp.achat.view.utils.UiUtils;

import java.util.LinkedList;
import java.util.List;

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.Holder> {

    public static final byte PAYLOAD_BUBBLE = 0;
    public static final byte PAYLOAD_READ = 1;

    public static final int TEXT_SIZE_SP = 14;
    public static final int EMOJI_SIZE_LARGEST_SP = 36;

    private final Context context;
    private List<Message> messages;
    private final int sp1;

    public MessageAdapter(Context context) {
        this.context = context;
        this.messages = new LinkedList<>();
        sp1 = UiUtils.sp2px(context, 1);
    }

    @Override
    public int getItemViewType(int position) {
        Message message = messages.get(position);
        return message.getTransferType() + message.getType();
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case Message.TRANSFER_TYPE_SEND + Message.TYPE_TEXT:
                return new TextHolder(LayoutInflater.from(context).inflate(R.layout.item_message_text_send, parent, false));
            case Message.TRANSFER_TYPE_RECEIVE + Message.TYPE_TEXT:
                return new TextHolder(LayoutInflater.from(context).inflate(R.layout.item_message_text_receive, parent, false));
            case Message.TRANSFER_TYPE_SEND + Message.TYPE_LOTTIE:
                return new LottieHolder(LayoutInflater.from(context).inflate(R.layout.item_message_lottie_send, parent, false));
            case Message.TRANSFER_TYPE_RECEIVE + Message.TYPE_LOTTIE:
                return new LottieHolder(LayoutInflater.from(context).inflate(R.layout.item_message_lottie_receive, parent, false));
            case Message.TRANSFER_TYPE_RECEIVE + Message.TYPE_DETAILS:
                return new DetailsHolder(LayoutInflater.from(context).inflate(R.layout.item_message_details, parent, false));
            case Message.TRANSFER_TYPE_RECEIVE + Message.TYPE_PROFILE:
                return new SingleProfileHolder(LayoutInflater.from(context).inflate(R.layout.item_message_profile, parent, false));
        }
        throw new RuntimeException("incorrect view type :" + getClass().getName());
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        holder.bind(messages.get(position));
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position, @NonNull List<Object> payloads) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position);
        } else {
            holder.bind(messages.get(position), payloads);
        }
    }

    @Override
    public void onViewRecycled(@NonNull Holder holder) {
        super.onViewRecycled(holder);
        if (holder instanceof LottieHolder) {
            ((LottieHolder) holder).recycle();
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    public void add(Message message) {
        boolean haveDateSeparatorPrev = false;
        if (message instanceof ChatMessage) {
            haveDateSeparatorPrev = setupMessagesBubble((ChatMessage) message);
        }
        if (haveDateSeparatorPrev) {
            Message detailsMessage = new DetailsMessage(message.getTimeMillis());
            messages.add(detailsMessage);
            messages.add(message);
            notifyItemRangeInserted(messages.size() - 2, 2);
        } else {
            messages.add(message);
            notifyItemInserted(messages.size() - 1);
        }
    }

    public void addAndScroll(List<Message> list) {
        add(message);
    }

    @SuppressLint("NotifyDataSetChanged")
    public void resetList(List<Message> list) {
        messages = list;
        notifyDataSetChanged();
    }

    private boolean setupMessagesBubble(ChatMessage message) {
        boolean haveDateSeparatorPrev = false;
        if (message.getTransferType() == Message.TRANSFER_TYPE_SEND) {
            if (messages.size() == 1) {
                message.setBubbleRes(MessageUtils.BUBBLE_RES_SEND_SINGLE);
                haveDateSeparatorPrev = true;
            } else {
                Message prev = messages.get(messages.size() - 1);
                if (prev instanceof ChatMessage
                        && prev.getTransferType() == Message.TRANSFER_TYPE_SEND
                        && message.getTimeMillis() - prev.getTimeMillis() < 60000
                ) {
                    message.setBubbleRes(MessageUtils.BUBBLE_RES_SEND_END);
                    if (messages.size() >= 3) {
                        Message prevPrev = messages.get(messages.size() - 2);
                        if (prevPrev instanceof ChatMessage
                                && prevPrev.getTransferType() == Message.TRANSFER_TYPE_SEND
                                && prev.getTimeMillis() - prevPrev.getTimeMillis() < 60000
                        ) {
                            ((ChatMessage) prev).setBubbleRes(MessageUtils.BUBBLE_RES_SEND_MIDDLE);
                        } else {
                            ((ChatMessage) prev).setBubbleRes(MessageUtils.BUBBLE_RES_SEND_START);
                        }
                    } else {
                        ((ChatMessage) prev).setBubbleRes(MessageUtils.BUBBLE_RES_SEND_START);
                    }
                    notifyItemChanged(messages.indexOf(prev), PAYLOAD_BUBBLE);
                } else {
                    if (!DateUtils.isToday(prev.getTimeMillis())) {
                        haveDateSeparatorPrev = true;
                    }
                    message.setBubbleRes(MessageUtils.BUBBLE_RES_SEND_SINGLE);
                }
            }
        } else {
            if (messages.size() == 1) {
                message.setBubbleRes(MessageUtils.BUBBLE_RES_RECEIVE_SINGLE);
                haveDateSeparatorPrev = true;
            } else {
                Message prev = messages.get(messages.size() - 1);
                if (prev instanceof ChatMessage
                        && prev.getTransferType() == Message.TRANSFER_TYPE_RECEIVE
                        && message.getTimeMillis() - prev.getTimeMillis() < 60000
                ) {
                    message.setBubbleRes(MessageUtils.BUBBLE_RES_RECEIVE_END);
                    if (messages.size() >= 3) {
                        Message prevPrev = messages.get(messages.size() - 2);
                        if (prevPrev instanceof ChatMessage
                                && prevPrev.getTransferType() == Message.TRANSFER_TYPE_RECEIVE
                                && prev.getTimeMillis() - prevPrev.getTimeMillis() < 60000
                        ) {
                            ((ChatMessage) prev).setBubbleRes(MessageUtils.BUBBLE_RES_RECEIVE_MIDDLE);
                        } else {
                            ((ChatMessage) prev).setBubbleRes(MessageUtils.BUBBLE_RES_RECEIVE_START);
                        }
                    } else {
                        ((ChatMessage) prev).setBubbleRes(MessageUtils.BUBBLE_RES_RECEIVE_START);
                    }
                    notifyItemChanged(messages.indexOf(prev), PAYLOAD_BUBBLE);
                } else {
                    if (!DateUtils.isToday(prev.getTimeMillis())) {
                        haveDateSeparatorPrev = true;
                    }
                    message.setBubbleRes(MessageUtils.BUBBLE_RES_RECEIVE_SINGLE);
                }
            }
        }
        return haveDateSeparatorPrev;
    }

    public abstract static class Holder extends RecyclerView.ViewHolder {

        public Holder(@NonNull View itemView) {
            super(itemView);
        }

        public abstract void bind(Message message);

        public void bind(Message message, List<Object> payloads) {

        }
    }

    public abstract static class ProfileHolder extends Holder implements View.OnClickListener {

        private final TextView name, description;

        public ProfileHolder(@NonNull View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.name);
            description = itemView.findViewById(R.id.description);
            itemView.setOnClickListener(this);
        }

        @Override
        public void bind(Message message) {
            ProfileMessage profileMessage = (ProfileMessage) message;
            name.setText(profileMessage.getContact().getName());
            description.setText(profileMessage.getContact().getBio());
        }

        @Override
        public void onClick(View v) {
            //todo go to profile page
        }
    }

    public static class SingleProfileHolder extends ProfileHolder {

        private final GroupAvatarView avatar;
        private final TextView rank, onlineTime;

        public SingleProfileHolder(@NonNull View itemView) {
            super(itemView);
            avatar = itemView.findViewById(R.id.avatar);
            rank = itemView.findViewById(R.id.rank);
            onlineTime = itemView.findViewById(R.id.lastOnline);
        }

        @Override
        public void bind(Message message) {
            super.bind(message);
            ProfileMessage profileMessage = (ProfileMessage) message;
            Contact contact = profileMessage.getContact();
            if (profileMessage.getContact().getType() == Contact.TYPE_SINGLE) {
                String[] avatars = contact.getAvatars();
                avatar.setAvatars(avatars.length > 0 ? avatars[0] : null);
                if (contact.getOnlineTime() == Contact.TIME_ONLINE) {
                    onlineTime.setText("");
                    onlineTime.setBackgroundResource(R.drawable.last_online_profile_bg_green);
                } else {
                    onlineTime.setText(TimeUtils.timeAgoShort(System.currentTimeMillis() - contact.getOnlineTime()));
                    onlineTime.setBackgroundResource(R.drawable.last_online_profile_bg_grey);
                }
                onlineTime.setVisibility(View.VISIBLE);
            } else {
                avatar.setAvatars(contact.getAvatars());
                onlineTime.setVisibility(View.GONE);
            }

            Pair<Integer, Integer> pair = PersonUtils.rankInt2rankStrResAndColor(contact.getRank());
            rank.setText(pair.first);
            rank.setTextColor(pair.second);
        }
    }

    public static class DetailsHolder extends Holder {

        public DetailsHolder(@NonNull View itemView) {
            super(itemView);
        }

        @Override
        public void bind(Message message) {
            ((TextView) itemView).setText(((DetailsMessage) message).getDetails());
        }
    }

    public abstract class ChatHolder extends Holder implements View.OnClickListener {

        protected final SimpleDraweeView avatar;
        protected final TextView time;
        protected final AppCompatImageView read;
        protected final View online;

        public ChatHolder(@NonNull View itemView) {
            super(itemView);
            avatar = itemView.findViewById(R.id.avatar);
            time = itemView.findViewById(R.id.time);
            read = itemView.findViewById(R.id.read);
            online = itemView.findViewById(R.id.lastOnline);

            itemView.setOnClickListener(this);
        }

        @Override
        public void bind(Message message) {
            ChatMessage chatMessage = (ChatMessage) message;
            setBubble(chatMessage);
            setRead(chatMessage);
            time.setText(chatMessage.getTime());
        }

        @Override
        public void bind(Message message, List<Object> payloads) {
            super.bind(message, payloads);
            for (Object payload : payloads) {
                switch ((byte) payload) {
                    case PAYLOAD_BUBBLE:
                        setBubble((ChatMessage) message);
                        break;
                    case PAYLOAD_READ:
                        setRead((ChatMessage) message);
                        break;
                }
            }
        }

        @SuppressLint("WrongConstant")
        protected void setBubble(ChatMessage message) {
            int bubbleRes = message.getBubbleRes();
            View bubbleView = getBubbleView();
            if (bubbleView != null) {
                bubbleView.setBackgroundResource(bubbleRes);
            }
            if (message.getTransferType() == Message.TRANSFER_TYPE_RECEIVE) {
                if (bubbleRes == MessageUtils.BUBBLE_RES_RECEIVE_SINGLE || bubbleRes == MessageUtils.BUBBLE_RES_RECEIVE_END) {
                    avatar.setVisibility(View.VISIBLE);
                    time.setVisibility(View.VISIBLE);

                    String[] avatars = message.getSender().getAvatars();
                    avatar.setImageURI(avatars.length > 0 ? avatars[0] : null);
                    online.setVisibility(message.getSender().getOnlineTime() == Contact.TIME_ONLINE ?
                            View.VISIBLE : View.GONE);

                } else {
                    avatar.setVisibility(View.GONE);
                    time.setVisibility(View.GONE);
                }
            } else if (message.getTransferType() == Message.TRANSFER_TYPE_SEND) {
                if (bubbleRes == MessageUtils.BUBBLE_RES_SEND_SINGLE || bubbleRes == MessageUtils.BUBBLE_RES_SEND_END) {
                    time.setVisibility(View.VISIBLE);
                } else {
                    time.setVisibility(View.GONE);
                }
            }
        }

        protected abstract View getBubbleView();

        private void setRead(ChatMessage chatMessage) {
            if (chatMessage.getTransferType() == Message.TRANSFER_TYPE_SEND) {
                read.setImageResource(chatMessage.getDeliveryRes());
            }
        }

        @Override
        public void onClick(View v) {
            int bubbleRes = ((ChatMessage) messages.get(getAdapterPosition())).getBubbleRes();
            if (bubbleRes != MessageUtils.BUBBLE_RES_RECEIVE_SINGLE
                    && bubbleRes != MessageUtils.BUBBLE_RES_RECEIVE_END
                    && bubbleRes != MessageUtils.BUBBLE_RES_SEND_SINGLE
                    && bubbleRes != MessageUtils.BUBBLE_RES_SEND_END
            ) {
                time.setVisibility(time.getVisibility() == View.GONE ? View.VISIBLE : View.GONE);
            }
        }
    }

    public class TextHolder extends ChatHolder {

        private final AXEmojiTextView textView;

        public TextHolder(@NonNull View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.text);
        }

        @Override
        public void bind(Message message) {
            super.bind(message);
            TextMessage textMessage = (TextMessage) message;
            String text = textMessage.getText();
            textView.setText(text);
            setTextSize(text, textMessage.getExtraTextSize());
        }

        @Override
        protected View getBubbleView() {
            return textView;
        }

        private void setTextSize(CharSequence text, int textSizeFactor) {
            float textSize = (TEXT_SIZE_SP + textSizeFactor) * sp1;
            float emojiSize = (TEXT_SIZE_SP + 3 + textSizeFactor) * sp1;
            boolean hasText = false;

            for (int i = 0; i < text.length(); i++) {
                if ((int) text.charAt(i) < 2000) {
                    hasText = true;
                    break;
                }
            }

            if (!hasText) {
                emojiSize = (EMOJI_SIZE_LARGEST_SP + textSizeFactor) * sp1;
                for (int i = 1; i < text.length(); i++) {
                    if (emojiSize <= (TEXT_SIZE_SP + 3 + textSizeFactor) * sp1) {
                        emojiSize = (TEXT_SIZE_SP + 3 + textSizeFactor) * sp1;
                        break;
                    }
                    emojiSize -= sp1;
                }
                textSize = emojiSize * 0.8f;
            }
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);
            textView.setEmojiSize((int) emojiSize);
        }
    }

    public class LottieHolder extends ChatHolder {

        private final AXrLottieImageView lottieImageView;

        public LottieHolder(@NonNull View itemView) {
            super(itemView);
            lottieImageView = itemView.findViewById(R.id.lottieImageView);
        }

        @Override
        public void bind(Message message) {
            super.bind(message);
            lottieImageView.setLottieDrawable(((LottieMessage) message).getDrawable());
            lottieImageView.playAnimation();
        }

        public void recycle() {
            lottieImageView.stopAnimation();
        }

        @Override
        protected View getBubbleView() {
            return null;
        }
    }
}

File: app/src/main/java/com/hyapp/achat/view/adapter/MessageAdapter.kt
package com.hyapp.achat.view.adapter

import android.content.Context
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.aghajari.rlottie.AXrLottieImageView
import com.facebook.drawee.view.SimpleDraweeView
import com.hyapp.achat.R
import com.hyapp.achat.model.entity.Contact
import com.hyapp.achat.model.entity.Message
import com.hyapp.achat.model.entity.UserConsts
import com.hyapp.achat.view.ProfileActivity
import com.hyapp.achat.view.component.GroupAvatarView
import com.hyapp.achat.view.component.emojiview.view.AXEmojiTextView
import com.hyapp.achat.view.utils.UiUtils
import com.hyapp.achat.viewmodel.utils.TimeUtils
import kotlinx.coroutines.ExperimentalCoroutinesApi


class MessageAdapter(val context: Context, val recyclerView: RecyclerView) :
    ListAdapter<Message, MessageAdapter.Holder>(DIFF_CALLBACK) {

    val sp1: Int = UiUtils.sp2px(context, 1F)
    var isLoadingMore = false
    lateinit var onLoadMore: () -> Unit
    var onListChanged: (() -> Unit)? = null

    companion object {
        const val PAYLOAD_BUBBLE: Byte = 0
        const val PAYLOAD_DELIVERY: Byte = 1
        const val PAYLOAD_ONLINE_TIME: Byte = 2
        const val PAYLOAD_BIO: Byte = 3

        val DIFF_CALLBACK: DiffUtil.ItemCallback<Message> =
            object : DiffUtil.ItemCallback<Message>() {
                override fun areItemsTheSame(oldItem: Message, newItem: Message): Boolean {
                    return oldItem.uid == newItem.uid
                }

                override fun areContentsTheSame(oldItem: Message, newItem: Message): Boolean {
                    return oldItem == newItem
                }

                override fun getChangePayload(oldItem: Message, newItem: Message): Any? {
                    return when {
                        oldItem.delivery != newItem.delivery -> {
                            PAYLOAD_DELIVERY
                        }
                        oldItem.time != newItem.time -> {
                            PAYLOAD_BUBBLE
                        }
                        oldItem.senderOnlineTime != newItem.senderOnlineTime -> {
                            PAYLOAD_ONLINE_TIME
                        }
                        oldItem.senderBio != newItem.senderBio -> {
                            PAYLOAD_BIO
                        }
                        else -> {
                            null
                        }
                    }
                }
            }
    }

    override fun submitList(list: MutableList<Message>?) {
        if (list != null) {
            super.submitList(ArrayList(list))
        }
    }

    fun getMessage(position: Int): Message {
        return getItem(position)
    }

    override fun onCurrentListChanged(
        previousList: MutableList<Message>,
        currentList: MutableList<Message>
    ) {
        onListChanged?.invoke()
    }

    override fun getItemViewType(position: Int): Int {
        val message = getItem(position)
        return message.transfer + message.type
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        when (viewType) {
            Message.TRANSFER_RECEIVE + Message.TYPE_TYPING -> return TypingHolder(
                LayoutInflater.from(context).inflate(R.layout.item_message_typing, parent, false)
            )
            Message.TRANSFER_SEND + Message.TYPE_TEXT -> return TextHolder(
                LayoutInflater.from(
                    context
                ).inflate(R.layout.item_message_text_send, parent, false)
            )
            Message.TRANSFER_RECEIVE + Message.TYPE_TEXT -> return TextHolder(
                LayoutInflater.from(
                    context
                ).inflate(R.layout.item_message_text_receive, parent, false)
            )
            Message.TRANSFER_SEND + Message.TYPE_LOTTIE -> return LottieHolder(
                LayoutInflater.from(
                    context
                ).inflate(R.layout.item_message_lottie_send, parent, false)
            )
            Message.TRANSFER_RECEIVE + Message.TYPE_LOTTIE -> return LottieHolder(
                LayoutInflater.from(
                    context
                ).inflate(R.layout.item_message_lottie_receive, parent, false)
            )
            Message.TRANSFER_RECEIVE + Message.TYPE_DETAILS -> return DetailsHolder(
                LayoutInflater.from(
                    context
                ).inflate(R.layout.item_message_details, parent, false)
            )
            Message.TRANSFER_RECEIVE + Message.TYPE_PROFILE -> return ProfileHolder(
                LayoutInflater.from(context).inflate(R.layout.item_message_profile, parent, false)
            )
        }
        throw RuntimeException("incorrect view type :" + javaClass.name)
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        holder.bind(getItem(position))
        if (position <= 0 && !isLoadingMore && itemCount > 2) {
            isLoadingMore = true
            if (::onLoadMore.isInitialized) {
                recyclerView.post { onLoadMore() }
            }
        }
    }

    override fun onBindViewHolder(holder: Holder, position: Int, payloads: List<Any?>) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position)
        } else {
            holder.bind(getItem(position), payloads)
        }
    }

    override fun onViewRecycled(holder: Holder) {
        super.onViewRecycled(holder)
        if (holder is LottieHolder) {
            holder.recycle()
        }
    }

//    fun add(list: MessageList, prevChanged: Boolean, addedCount: Int) {
//        messages = list
//        val size = messages.size
//        if (prevChanged)
//            notifyItemChanged(size - 1 - addedCount, PAYLOAD_BUBBLE)
//        notifyItemRangeInserted(size - addedCount, addedCount)
//    }
//
//    fun addPaging(list: MessageList, addedCount: Int, firstChanged: Boolean) {
//        messages = list
//        if (firstChanged)
//            notifyItemChanged(0, PAYLOAD_BUBBLE)
//        notifyItemRangeInserted(0, addedCount)
//    }

    abstract inner class Holder(view: View) : RecyclerView.ViewHolder(view) {
        abstract fun bind(message: Message)

        open fun bind(message: Message, payloads: List<Any?>) {}
    }

    @ExperimentalCoroutinesApi
    @Suppress("LeakingThis")
    inner class ProfileHolder(itemView: View) : Holder(itemView), View.OnClickListener {
        private val name: TextView = itemView.findViewById(R.id.name)
        private val bio: TextView = itemView.findViewById(R.id.description)
        private val avatar: GroupAvatarView = itemView.findViewById(R.id.avatar)
        private val rank: TextView = itemView.findViewById(R.id.rank)
        private val onlineTime: TextView = itemView.findViewById(R.id.lastOnline)

        override fun bind(message: Message) {
            name.text = message.senderName
            bio.text = message.senderBio
            val contact = message.getContact()
            if (message.isPv) {
                avatar.setAvatars(contact.avatars)
                setOnlineTime(message.senderOnlineTime)
                onlineTime.visibility = View.VISIBLE
            } else {
                avatar.setAvatars(contact.avatars)
                onlineTime.visibility = View.GONE
            }
            val pair = UserConsts.rankInt2rankStrResAndColor(contact.rank)
            rank.setText(pair.first)
            rank.setTextColor(pair.second)
        }

        override fun bind(message: Message, payloads: List<Any?>) {
            super.bind(message, payloads)
            for (payload in payloads) {
                when (payload as Byte) {
                    PAYLOAD_ONLINE_TIME -> setOnlineTime(message.senderOnlineTime)
                    PAYLOAD_BIO -> bio.text = message.senderBio

                }
            }
        }

        private fun setOnlineTime(time: Long) {
            if (time == Contact.TIME_ONLINE) {
                onlineTime.text = ""
                onlineTime.setBackgroundResource(R.drawable.last_online_profile_bg_green)
            } else {
                onlineTime.text =
                    TimeUtils.timeAgoShort(System.currentTimeMillis() - time)
                onlineTime.setBackgroundResource(R.drawable.last_online_profile_bg_grey)
            }
        }

        override fun onClick(v: View) {
            val message = getItem(adapterPosition)
            ProfileActivity.start(context, message.getContact().getUser())
        }

        init {
            itemView.setOnClickListener(this)
        }
    }

    inner class DetailsHolder(itemView: View) : Holder(itemView) {
        override fun bind(message: Message) {
            (itemView as TextView).text = message.text
        }
    }

    @Suppress("LeakingThis")
    abstract inner class ChatHolder(itemView: View) : Holder(itemView), View.OnClickListener {
        protected val avatar: SimpleDraweeView? = itemView.findViewById(R.id.avatar)
        protected val time: TextView = itemView.findViewById(R.id.time)
        protected val delivery: AppCompatImageView? = itemView.findViewById(R.id.read)

        override fun bind(message: Message) {
            setBubble(message)
            setDelivery(message)
            time.text = TimeUtils.millis2DayTime(message.time)
        }

        override fun bind(message: Message, payloads: List<Any?>) {
            super.bind(message, payloads)
            for (payload in payloads) {
                when (payload as Byte) {
                    PAYLOAD_BUBBLE -> setBubble(message)
                    PAYLOAD_DELIVERY -> setDelivery(message)
                }
            }
        }

        private fun setDelivery(message: Message) {
            if (message.transfer == Message.TRANSFER_SEND) {
                when (message.delivery) {
                    Message.DELIVERY_READ -> delivery?.setImageResource(R.drawable.msg_read_contact)
                    Message.DELIVERY_SENT -> delivery?.setImageResource(R.drawable.msg_unread_contact)
                    Message.DELIVERY_WAITING -> delivery?.setImageResource(R.drawable.msg_waiting_contact)
                }
            }
        }

        private fun setBubble(message: Message) {
            val bubble: Byte = message.bubble
            val bubbleView = bubbleView
            bubbleView?.let {
                if (message.transfer == Message.TRANSFER_SEND) {
                    when (bubble) {
                        Message.BUBBLE_END -> it.setBackgroundResource(R.drawable.chat_bubble_send_end)
                        Message.BUBBLE_MIDDLE -> it.setBackgroundResource(R.drawable.chat_bubble_send_middle)
                        Message.BUBBLE_START -> it.setBackgroundResource(R.drawable.chat_bubble_send_start)
                        Message.BUBBLE_SINGLE -> it.setBackgroundResource(R.drawable.chat_bubble_send_single)
                    }
                    if (bubble == Message.BUBBLE_SINGLE || bubble == Message.BUBBLE_END) {
                        time.visibility = View.VISIBLE
                    } else {
                        time.visibility = View.GONE
                    }
                } else {
                    when (bubble) {
                        Message.BUBBLE_END -> it.setBackgroundResource(R.drawable.chat_bubble_receive_end)
                        Message.BUBBLE_MIDDLE -> it.setBackgroundResource(R.drawable.chat_bubble_receive_middle)
                        Message.BUBBLE_START -> it.setBackgroundResource(R.drawable.chat_bubble_receive_start)
                        Message.BUBBLE_SINGLE -> it.setBackgroundResource(R.drawable.chat_bubble_receive_single)
                    }
                    if (bubble == Message.BUBBLE_SINGLE || bubble == Message.BUBBLE_END) {
                        avatar?.visibility = View.VISIBLE
                        time.visibility = View.VISIBLE
                        avatar?.setImageURI(message.firstSenderAvatar)
                    } else {
                        avatar?.visibility = View.GONE
                        time.visibility = View.GONE
                    }
                }
            }
        }

        protected abstract val bubbleView: View?

        override fun onClick(v: View) {
            val bubble: Byte = getItem(adapterPosition).bubble
            if (bubble != Message.BUBBLE_SINGLE && bubble != Message.BUBBLE_END && bubble != Message.BUBBLE_SINGLE && bubble != Message.BUBBLE_END) {
                time.visibility = if (time.visibility == View.GONE) View.VISIBLE else View.GONE
            }
        }

        init {
            itemView.setOnClickListener(this)
        }
    }

    inner class TextHolder(itemView: View) : ChatHolder(itemView) {
        private val textView: AXEmojiTextView = itemView.findViewById(R.id.text)
        override fun bind(message: Message) {
            super.bind(message)
            val text = message.text
            textView.text = text
            val sizes = message.setAndGetTextSizes(sp1)
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, sizes.first)
            textView.setEmojiSize(sizes.second)
        }

        override val bubbleView: View
            get() = textView

    }

    inner class LottieHolder(itemView: View) : ChatHolder(itemView) {
        private val lottieImageView: AXrLottieImageView =
            itemView.findViewById(R.id.lottieImageView)

        override fun bind(message: Message) {
            super.bind(message)
//            lottieImageView.lottieDrawable = (message as LottieMessage).drawable
//            lottieImageView.playAnimation()
        }

        fun recycle() {
//            lottieImageView.stopAnimation()
        }

        override val bubbleView: View?
            get() = null

    }

    inner class TypingHolder(itemView: View) : Holder(itemView) {
        val avatar = itemView.findViewById<SimpleDraweeView>(R.id.avatar)

        override fun bind(message: Message) {
            val ava = message.senderAvatars
            avatar.setImageURI(if (ava.isNotEmpty()) ava[0] else null)
        }
    }
}